src/crimson/os/seastore/seastore.cc

```diff
@@ -400,23 +400,127 @@ seastar::future<store_statfs_t> SeaStore::stat() const
+      return collection_manager->list(coll_root, t);
+    }).si_then([ch](auto colls) {
+      auto it = std::find_if(colls.begin(), colls.end(),
+        [ch](const std::pair<coll_t, coll_info_t>& element) {
+          return element.first == ch->get_cid();
+      });
+      if (it != colls.end()) {
+        return TransactionManager::read_extent_iertr::make_ready_future<
+          std::optional<unsigned>>(it->second.split_bits);
+      } else {
+        return TransactionManager::read_extent_iertr::make_ready_future<
+	  std::optional<unsigned>>(std::nullopt);
+      }
+    });
+}
+
+OnodeManager::RangeTuple
+SeaStore::get_objs_range(CollectionRef ch, unsigned bits) const
+{
+  ghobject_t coll_range_temp_start, coll_range_temp_end;
+  ghobject_t coll_range_start, coll_range_end;
+  get_coll_range(ch->get_cid(),
+                 bits,
+                 &coll_range_temp_start,
+                 &coll_range_temp_end,
+                 &coll_range_start,
+                 &coll_range_end);
+  return std::make_tuple(coll_range_temp_start,
+                         coll_range_temp_end,
+                         coll_range_start,
+                         coll_range_end);
+}
+
 seastar::future<std::tuple<std::vector<ghobject_t>, ghobject_t>>
 SeaStore::list_objects(CollectionRef ch,
-                        const ghobject_t& start,
-                        const ghobject_t& end,
-                        uint64_t limit) const
+                       const ghobject_t start,
+                       const ghobject_t end,
+                       uint64_t limit) const
 {
   using RetType = typename OnodeManager::list_onodes_bare_ret;
+  if (start.is_max() || start.hobj.is_max()) {
+    return seastar::make_ready_future<
+      std::tuple<std::vector<ghobject_t>, ghobject_t>>(
+      std::make_tuple(std::vector<ghobject_t>(), ghobject_t::get_max()));
+  }
   return seastar::do_with(
-      RetType(),
-      [this, start, end, limit] (auto& ret) {
-    return repeat_eagain([this, start, end, limit, &ret] {
+    RetType(),
+    [this, ch, start, end, limit] (auto& ret) {
+    return repeat_eagain([this, ch, start, end, limit, &ret] {
       return transaction_manager->with_transaction_intr(
         Transaction::src_t::READ,
         "list_objects",
-        [this, start, end, limit](auto &t)
+        [this, ch, start, end, limit](auto &t)
       {
-        return onode_manager->list_onodes(t, start, end, limit);
+        return get_coll_bits(ch, t)
+          .si_then([this, ch, &t, start, end, limit](auto bits) {
+          if (!bits) {
+            return OnodeManager::list_onodes_iertr::make_ready_future<
+	      OnodeManager::list_onodes_bare_ret>(std::make_tuple(
+	      std::vector<ghobject_t>(), ghobject_t::get_max()));
+          } else {
+            auto filter = get_objs_range(ch, *bits);
+            return onode_manager->list_onodes_with_filter(
+              t, start, end, limit, ch->get_cid().get_min_hobj(), filter);
```

Comment:
I don't think list_onodes_with_filter is actually necessary.  We should be able to simply invoke list_onodes possibly twice if the range crosses the boundary.

Reply:
yes, just invoke list_onodes twice.

Reply:
@athanatos please review again.